docs/Workflow.md

```diff
@@ -51,102 +95,180 @@ git remote set-url --push upstream no_push
 git remote -v
 ```
 
-### Step 3: Branch
+## Step 3: Understanding the repository structure
 
-Get your local main up to date:
+At the end of the repository setup commands above:
+- `upstream` will refer to the original ONNX-MLIR repository.
+- `origin` will refer to your own fork of ONNX-MLIR.
 
+Never commit anything to your fork's `main` branch, the only way you should update `main` is from `upstream`. The procedure to update your fork's `main` branch is listed in Step 4.
+
+## Step 4: Keeping your repository up to date
+
+To keep your ONNX-MLIR fork's `main` up to date perform the following steps:
+
+1. Fetch the latest versions of your fork (`origin`) and the `upstream` repositories:
+```sh
+git fetch --all
+```
+
+2. Update the `main` branch on your fork:
 ```sh
-cd $working_dir/onnx-mlir
-git fetch upstream
 git checkout main
-git rebase upstream/main
+git pull
+git merge upstream/main
+git push origin main
 ```
 
-Branch from main:
+Provided you have never committed anything to your fork's `main` branch directly, all the updates to your fork's `main` should be fast forwards.
 
+3. The `main` branch of your fork should now be identical to the `main` branch of `upstream`. To check you can do:
 ```sh
-git checkout -b myfeature
+git diff upstream/main
 ```
+and the command will return immediately signaling that no differences exist between `upstream/main` and `origin/main`
 
-### Step 4: Develop
+4. (Optional) Assuming your `main` is up to date, to update any branches you are currently working on to use the latest ONNX-MLIR, you need to do the following:
 
-#### Edit the code
+```sh
+git checkout my-branch
+git merge origin/main
```

Comment:
Personally, I find this command dangerous. Most important is to have upstream/main -> local main. Then if we want, we can move local main into origin/main, I do it, but this is not key. So I find it safer to mention here only a `git merge main` because it gets it from the local, which is guaranteed to be updated if you did a fetch/merge from upstream. The command specifying `origin/main` is only good if the user did the additional step to push it to its origin.

Reply:
The update instructions include the push command hence the origin/main here so it all works. The key is that all important branches like main should always be pushed, there's no reason not to have them updated on GitHub also.

Reply:
I also like the lack of ambiguity, you know exactly which main you're brining in.